### ISSUE TITLE
Iterating over large folder collection in a paged way still throws ThrottledException

### ISSUE THREAD

**Issue by Ironbell** [pnp/pnpcore#797]
## Category

- [x] Bug

## Describe the bug

We do have a folder with more than 5000 subfolders. In the documentation, it states that iterating the IFolderCollection using await foreach would take care of paging - but we still run into the ThrottledException immediately. We also tried paging using CAML queries, but even with paging as shown on your documentation, we immediately run into the Exception as well.

```
using var context = await this.pnpContextFactory.CreateAsync("ConnectionName");
  var rootFolder = (await context.Web.Lists.GetByTitleAsync("LibraryName", p => p.RootFolder)).RootFolder;

  await foreach (var folder in rootFolder .Folders)
  {
       // it already throws an exception in the await foreach
  }
```

## Steps to reproduce

1. Have a document library with more than 5000 folders in the root folder
2. Try the above code
3. Observe the throttled exception

## Expected behavior

I expect the await foreach to do the paging internally as stated in the documentation.

**Comment by jansenbe** [comment]
@Ironbell : can you share a stacktrace of the exception you see?

**Comment by jansenbe** [comment]
Also, does https://github.com/pnp/pnpcore/issues/793 feel related? Maybe it's the same exception you're getting?

**Comment by Ironbell** [comment]
Hi @jansenbe thanks for your quick answer. #793 is related in a way, we also had this problem with EnsureFolderExistsAsync, as this method scales with the number of folders in there. For the EnsureFolderExistsAsync, I've made my own implementation though that is O(1) instead of O(n), n = number of folders and very fast:

```
private static async Task<IFolder> EnsureFolderAsync(string directoryPath, IFolder rootFolder)
        {
            var currentFolder = rootFolder;

            var childFolderNames = directoryPath.Split(new[] { '/' }, StringSplitOptions.RemoveEmptyEntries);

            var context = rootFolder.PnPContext;

            await rootFolder.LoadAsync(p => p.ServerRelativeUrl);

            var currentUrl = rootFolder.ServerRelativeUrl;
            var currentFolderWasCreated = false;

            foreach (var folderName in childFolderNames)
            {
                currentUrl = $"{currentUrl}/{folderName}";

                if (!currentFolderWasCreated)
                {
                    try
                    {
                        var nextFolder = await context.Web.GetFolderByServerRelativeUrlAsync(currentUrl);

                        currentFolder = nextFolder;
                    }
                    catch (PnP.Core.SharePointRestServiceException)
                    {
                        currentFolderWasCreated = true;
                        currentFolder = await currentFolder.AddFolderAsync(folderName);
                    }
                }
                else
                {
                    currentFolder = await currentFolder.AddFolderAsync(folderName);
                }
            }

            return currentFolder;
        }
```

Basically just getting it via the relative Url instead of iterating over all the folders. (and catching the exception if it doesn't find it)

But for our other issue, it still persists as we really have to iterate over all folders. The stack trace is the following:

Message: The attempted operation is prohibited because it exceeds the list view threshold.

![image](https://user-images.githubusercontent.com/9303244/159703265-2d37239d-87f7-40b3-bf64-9da4c48aa69a.png)

Stack trace
```
PnP.Core.SharePointRestServiceException
  HResult=0x80131500
  Message=SharePoint Rest service exception
  Source=PnP.Core
  StackTrace:
   at PnP.Core.Services.BatchClient.<ExecuteSharePointRestInteractiveAsync>d__59.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PnP.Core.Services.BatchClient.<ExecuteSharePointRestBatchAsync>d__51.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PnP.Core.Services.BatchClient.<ExecuteBatch>d__36.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at PnP.Core.Services.PnPContext.<ExecuteAsync>d__77.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at PnP.Core.QueryModel.DataModelQueryService`1.<ExecuteQueryAsync>d__16.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at PnP.Core.QueryModel.BaseQueryProvider.<GetAsyncEnumerable>d__11`1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.ThrowForFailedGetResult(Int16 token)
   at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.GetResult(Int16 token)
   at PnP.Core.QueryModel.BaseQueryProvider.<GetAsyncEnumerable>d__11`1.System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)
   at System.Runtime.CompilerServices.ValueTaskAwaiter`1.GetResult()
   at Helion.Lead.Lead.Repository.LeadDocumentAndPictureSharePointRepository.<SearchForLeadAddressFolderAsync>d__42.MoveNext() in C:\Git\helion-portale\Source\Helion.Lead\Lead\Repository\LeadDocumentAndPictureSharePointRepository.cs:line 906
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Helion.Lead.Lead.Repository.LeadDocumentAndPictureSharePointRepository.<SearchForLeadAddressFolderAsync>d__42.MoveNext() in C:\Git\helion-portale\Source\Helion.Lead\Lead\Repository\LeadDocumentAndPictureSharePointRepository.cs:line 906
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Helion.Lead.Lead.Repository.LeadDocumentAndPictureRepository.<TriggerManualDocumentAndPicturesFileSyncAsync>d__18.MoveNext() in C:\Git\helion-portale\Source\Helion.Lead\Lead\Repository\LeadDocumentAndPictureRepository.cs:line 233
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Helion.Api.PC.AzureFileShare.LeadDocumentAndPictureApplicationService.<TriggerManualDocumentAndPicturesFileSyncAsync>d__20.MoveNext() in C:\Git\helion-portale\Source\Helion.Api.PCCore\AzureFileShare\LeadDocumentAndPictureApplicationService.cs:line 318

  This exception was originally thrown at this call stack:
    [External Code]
    Helion.Lead.Lead.Repository.LeadDocumentAndPictureSharePointRepository.SearchForLeadAddressFolderAsync(Helion.Lead.Lead.Domain.Lead, Helion.Mandator.Mandator.Domain.Mandator) in LeadDocumentAndPictureSharePointRepository.cs
    [External Code]
    Helion.Lead.Lead.Repository.LeadDocumentAndPictureSharePointRepository.SearchForLeadAddressFolderAsync(Helion.Lead.Lead.Domain.Lead, Helion.Mandator.Mandator.Domain.Mandator) in LeadDocumentAndPictureSharePointRepository.cs
    [External Code]
    Helion.Lead.Lead.Repository.LeadDocumentAndPictureRepository.TriggerManualDocumentAndPicturesFileSyncAsync(Helion.Lead.Lead.Domain.Lead) in LeadDocumentAndPictureRepository.cs
    [External Code]
    Helion.Api.PC.AzureFileShare.LeadDocumentAndPictureApplicationService.TriggerManualDocumentAndPicturesFileSyncAsync(int, string) in LeadDocumentAndPictureApplicationService.cs
```

**Comment by jansenbe** [comment]
@Ironbell : I'll do some testing and provide feedback. The implicit paging you refer to works for list items but as you're enumerating folders things are different. Using a CAML based approach should work however, but I'll test with folders to ensure it does.

**Comment by jansenbe** [comment]
@Ironbell : below code allows me to iterate over a library with 5500+ folders inside a sub folder:

```csharp
var library = await context.Web.Lists.GetByTitleAsync("PNP_SDK_TEST_LargeFolderLibrary");

await foreach (var listItem in library.Items.QueryProperties(p => p.Folder, p => p.All).AsAsyncEnumerable())
{
    // Do something with the list item
    if (listItem["ContentTypeId"].ToString().StartsWith("0x0120"))
    {
        // take action
        Debug.WriteLine(listItem.Folder.ServerRelativeUrl);
    }
}
```

**Comment by jansenbe** [comment]
@Ironbell : paging over the Folders collection in REST will never work: `_api/Web/getFolderById('e504859d-8d7b-4650-b63f-641508ddd5f8')/Folders?$top=100` will always throw the error you see, hence you need to use the "page over the list items" approach as shown above.

If this unblocks you, then feel free to close the issue.

**Comment by jansenbe** [comment]
For future needs, here's the test code used to create the library with 5500+ folders:

```csharp
var listTitle = TestCommon.GetPnPSdkTestAssetName("LargeFolderLibrary");
var list = await context.Web.Lists.AddAsync(listTitle, ListTemplateType.DocumentLibrary);
await list.EnsurePropertiesAsync(l => l.RootFolder);

var folder = await list.RootFolder.AddFolderAsync("Level1");

int j = 0;
var batch = context.NewBatch();
for (int i = 1; i < 5555; i++)
{
    await folder.Folders.AddBatchAsync(batch, $"Level1-{i}");
    j++;

    if (j == 20)
    {
        j = 0;
        await context.ExecuteAsync(batch);
        batch = context.NewBatch();
    }
}
```

**Comment by Ironbell** [comment]
Thank you @jansenbe for your reply. I actually tried iterating over the list items instead, but I get stuck on the second item. I set the `DefaultPageSize` down to 1 in the configuration so I should be able to observe every item being loaded independently. It loads the first folder and then just seems to go into an endless loop. I'm having a hard time to debug what's happening there, so I'm attaching a gif of the debugging session, maybe you know what could be the problem.

![asyncenumerable](https://user-images.githubusercontent.com/9303244/159802636-06e4e2b0-100b-4c7d-8100-afc3a9579fa6.gif)

**Comment by jansenbe** [comment]
Can you try to just let the code run, after the first page it will load all other pages before returning control. A fiddler trace should show that...we'll need to investigate why the code is not switching  into the loop body per page

**Comment by Ironbell** [comment]
@jansenbe  just downloaded fiddler, I'll let you know in here about the calls.

I'm not planning to stay with the page size of 2 of course, we need bigger chunks and less calls in the end. but it is nice for debugging right now :)

So this is with the page size of 2, returning 2 folders
![image](https://user-images.githubusercontent.com/9303244/159920938-c7dea552-7ffb-4df9-b90b-6a81c2369517.png)

I got the entire archive of the two sharepoint calls it does in the first iteration, would you want me to send it to you via email? I can't attach it here unfortunately

![image](https://user-images.githubusercontent.com/9303244/159921134-6bb45cd5-c985-46bb-9d06-bb30babdd422.png)

**Comment by jansenbe** [comment]
@Ironbell : does this now work for you, meaning when you let the code retrieve all pages you should end up in your for loop code  for page 2 to n, page 1 is returned immediately.

**Comment by Ironbell** [comment]
@jansenbe, yes, page 1 is returned immediately but page 2 never is, it hangs in the loop after retrieving page 1

**Comment by Ironbell** [comment]
@jansenbe it seems to hang inside the async operation within PnPCore afterwards and I can post the stack trace here of the Task
![image](https://user-images.githubusercontent.com/9303244/160002159-db2fd95f-3524-47fb-9690-bf99c334645f.png)

**Comment by jansenbe** [comment]
I assume you're using the last nightly @Ironbell ?

**Comment by Ironbell** [comment]
@jansenbe I am using the latest stable 1.5.0 via nuget package manager. Ill see whether I can use the nightly preview package instead. Worst case 'll get the source and investigate myself where it hangs. I also tried looping over the items in a synchronous way and wouldn't even get the first page - and the CAML queries I tried would all run into the Throttled exception, even when I did the paging as described in the documentation. I will let you know how it goes with the preview packages.

**Comment by jansenbe** [comment]
@Ironbell : that explains, we fixed #701 after shipping 1.5. Actually I'm planning on shipping 1.6 today/early next week, so you also wait for that for your testing?

**Comment by Ironbell** [comment]
@jansenbe I updated to 1.5.100-nightly on nuget and still have the same behavior (stuck on the second page with async and stuck on the first page without async). I'll download the source now and debug in there, I'll let you know when I know more. We'll get there.

**Comment by jansenbe** [comment]
@Ironbell , yes please try to debug. Do note that we support source link, so you don't have to pull down the code and built your own version for debugging. See https://pnp.github.io/pnpcore/using-the-sdk/basics-debug.html#configuring-visual-studio-to-support-debugging-pnp-core-sdk-packages-via-source-link for instructions on how to use this.

Also, the release will have to be postponed as I just noticed our code signing cert renewal got stuck, so first need to resolve that.

**Comment by jansenbe** [comment]
@Ironbell : are you using .NET 5/6/Core or .NET Framework?

**Comment by Ironbell** [comment]
@jansenbe we're using NET 6.0. Thanks for the instructions for the source link!

**Comment by Ironbell** [comment]
@jansenbe okay everything is alright. I got the source and found out why it seems to hang:

It first gets the first page and starts immediately iterating on what it received. So in the image below, it hits the first breakpoint once and then the second one for number of paged items times.

But after that, it goes into a loop where it fetches the items page for page with the nextlink and loads them all into memory. this takes a while of course. Only after all of it is loaded, page by page, it starts iterating on the rest. I assumed that every time a page is returned, it would go and iterate again on the items it received - I didn't realize it does page 1 and then page 2-n in one batch.

![image](https://user-images.githubusercontent.com/9303244/160106927-382e434e-c5ef-4e12-aa80-10e60f62b43f.png)

I will close this issue now, with the 1.5.100-nightly it terminates after a while. I'll wait for 1.6.0 then to go live. Thanks for all your work and sorry for the troubles.

**Comment by Ironbell** [comment]
It might still be a good idea, memory wise, to load page after page and iterate on the page, then clear the page and then load the next page, as page 2-n might be quite huge to store in memory.

**Comment by ChrTall** [comment]
@jansenbe 
I have the same problem as Ironbell. When I use implicit paging over Folders I hit the query limit. When I try to access the Folder Property to execute MoveToAsync() It does not work, because the following exception is thrown:
 at System.String.Substring(Int32 startIndex, Int32 length)
         at PnP.Core.Services.BatchClient.SharePointRestBatchSplitting(Batch batch)
         at PnP.Core.Services.BatchClient.ExecuteSharePointRestBatchAsync(Batch batch)
         at PnP.Core.Services.BatchClient.ExecuteBatch(Batch batch)
         at PnP.Core.Model.BaseDataModel`1.RequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
         at PnP.Core.Model.BaseDataModel`1.RawRequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
         at PnP.Core.Model.SharePoint.Folder.MoveToAsync(String destinationUrl, MoveCopyOptions options)
When I try to move Folders by directly loading the Folder and not over the List item it works. But only for  libraries with less than 5000 items.
